commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2023 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.tiered.storage.commons.chunkindex;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * The chunk index for transformed chunks of variable size.
+ *
+ * <p>The most important quality is that while the original file chunks
+ * are of the same size (apart from maybe the final one), the transformed chunks
+ * have different sizes that must be specified and stored explicitly.
+ * For example, this is a result of compression.
+ *
+ * <p>An example:
+ * <pre>
+ * Original file:        |  Transformed file:
+ * [0-100)   - size 100  |  [0-30)  - size 30
+ * [100-200) - size 100  |  [30-50) - size 20
+ * [200-250) - size 50   |  [50-60) - size 10
+ * </pre>
+ */
+public class VariableSizeChunkIndex extends AbstractChunkIndex {
+    private final List<Integer> transformedChunks;
+
+    // This only a materialization for convenience and performance,
+    // it should not be persisted.
+    private final List<Chunk> chunks;
+
+    public VariableSizeChunkIndex(final int originalChunkSize,
+                                  final int originalFileSize,
+                                  final List<Integer> transformedChunks) {
+        super(originalChunkSize, originalFileSize, transformedChunks.get(transformedChunks.size() - 1));
+        this.transformedChunks = List.copyOf(Objects.requireNonNull(transformedChunks));
```

Comment:
Also I think builder is not reusable, right? So it perhaps does not make sense to make a copy here since the constructor is called only in builder's `finish0`

Reply:
That's true, the builder is one-time. However, the same reason as for keeping `Objects.requireNonNull`

Reply:
As you wish but I do not see any value neither in copying it here nor making it unmodifiable since it's a private field.

Reply:
Well, you're right. I removed the copying